wasm-rpc-stubgen/src/commands/declarative.rs

```diff
@@ -0,0 +1,302 @@
+};
+use crate::stub::StubDefinition;
+use crate::{commands, WasmRpcOverride};
+use anyhow::{anyhow, Context};
+use colored::Colorize;
+use glob::glob;
+use itertools::Itertools;
+use std::io::Write;
+use std::path::PathBuf;
+use tempfile::TempDir;
+
+pub struct Config {
+    pub app_resolve_mode: ApplicationResolveMode,
+    pub skip_up_to_date_checks: bool,
+}
+
+pub enum ApplicationResolveMode {
+    Automatic,
+    Explicit(Vec<PathBuf>),
+}
+
+pub fn init(component_name: String) -> anyhow::Result<()> {
+    let file_name = DEFAULT_CONFIG_FILE_NAME;
+
+    let mut file = std::fs::File::create_new(file_name)
+        .with_context(|| format!("Failed to create {}", file_name))?;
+
+    let app = init_oam_app(component_name);
+    file.write_all(app.to_yaml_string().as_bytes())
+        .with_context(|| format!("Failed to write {}", file_name))?;
+
+    Ok(())
+}
+
+pub async fn pre_build(config: Config) -> anyhow::Result<()> {
+    let app = to_anyhow(
+        "Failed to load application manifest(s), see problems above".to_string(),
+        load_app(config.app_resolve_mode),
+    )?;
+
+    for component_name in app.all_wasm_rpc_dependencies() {
+        log_action("Building", format!("component stub: {}", component_name));
+
+        let target_root = TempDir::new()?;
+        let canonical_target_root = target_root.path().canonicalize()?;
+
+        let stub_def = StubDefinition::new(
+            &app.stub_source_wit_root(&component_name),
+            &canonical_target_root,
+            &app.stub_world(&component_name),
+            &app.stub_crate_version(&component_name),
+            &WasmRpcOverride {
+                wasm_rpc_path_override: app.stub_wasm_rpc_path(&component_name),
+                wasm_rpc_version_override: app.stub_wasm_rpc_version(&component_name),
+            },
+            app.stub_always_inline_types(&component_name),
+        )
+        .context("Failed to gather information for the stub generator")?;
+
+        commands::generate::build(
+            &stub_def,
+            &app.stub_wasm(&component_name),
+            &app.stub_wit(&component_name),
+        )
+        .await?
+    }
+
+    for (component_name, component) in &app.components_by_name {
+        if !component.wasm_rpc_dependencies.is_empty() {
+            log_action(
+                "Adding",
+                format!("stub wit dependencies to {}", component_name),
+            )
+        }
+
+        for dep_component_name in &component.wasm_rpc_dependencies {
+            log_action(
+                "Adding",
+                format!(
+                    "{} stub wit dependency to {}",
+                    dep_component_name, component_name
+                ),
+            );
+
+            commands::dependencies::add_stub_dependency(
+                &app.stub_wit(dep_component_name),
+                &app.component_wit(component_name),
+                true,  // TODO: is overwrite ever not needed?
```

Comment:
I added it to protect the potentially user-written files from getting overwritten. I think in the declarative mode, especially if we do the next planned step which moves the user-defined part "inside", we can assume that all wit directories are managed by this tool so it can freely modify everything except the original user defined wit specs.

Reply:
Added a note about always using true for this in declarative mode